File: src/aria.mjs
// Shared ARIA role + accessible-name helpers (used by extract, ax, and locator).
// Pragmatic, no-layout heuristics — enough for getByRole/getByText resolution and
// the agent view.

import { textOf } from "./dom-ops.mjs";

// Implicit ARIA role for common interactive/structural tags.
const IMPLICIT_ROLE = {
  a: "link",
  button: "button",
  select: "combobox",
  textarea: "textbox",
};

// Implicit role for <input> keyed by its `type` (default → textbox).
const INPUT_ROLE = {
  checkbox: "checkbox",
  radio: "radio",
  button: "button",
  submit: "button",
  reset: "button",
};

export function implicitRole(tag, type) {
  if (tag === "input") return INPUT_ROLE[type] ?? "textbox";
  return IMPLICIT_ROLE[tag] ?? "generic";
}

/** Resolved role: explicit `role` attribute, else the implicit tag/type role. */
export function roleOf(el) {
  const explicit = el.getAttribute("role");
  if (explicit) return explicit;
  return implicitRole(el.tagName.toLowerCase(), el.getAttribute("type")?.toLowerCase());
}

// First trimmed, non-empty string produced by one of the candidate getters.
function firstNonEmpty(getters) {
  for (const get of getters) {
    const v = get();
    const t = v == null ? "" : v.trim();
    if (t) return t;
  }
  return "";
}

// Accessible name, cheap heuristic: aria-label > text > placeholder > value > title.
export function accessibleName(el) {
  return firstNonEmpty([
    () => el.getAttribute("aria-label"),
    () => el.textContent,
    () => el.getAttribute("placeholder"),
    () => el.getAttribute("value"),
    () => el.getAttribute("title"),
  ]);
}

// IDREF list of an attribute (e.g. aria-describedby="a b"), filtered to non-empty.
function idList(el, attr) {
  return (el.getAttribute(attr) ?? "").split(/\s+/).filter(Boolean);
}

// Concatenated trimmed text of the referenced elements, space-joined.
function resolveIds(doc, ids) {
  const els = ids.map((id) => doc.getElementById(id)).filter(Boolean);
  return els
    .map((e) => textOf(e))
    .filter(Boolean)
    .join(" ");
}

// Accessible description: aria-describedby targets, else the `title` attribute.
export function accessibleDescription(el) {
  const ids = idList(el, "aria-describedby");
  if (ids.length) return resolveIds(el.ownerDocument, ids);
  return (el.getAttribute("title") ?? "").trim();
}

// Accessible error message: aria-errormessage targets, only when aria-invalid.
export function accessibleErrorMessage(el) {
  if (el.getAttribute("aria-invalid") !== "true") return "";
  return resolveIds(el.ownerDocument, idList(el, "aria-errormessage"));
}

File: rust/crates/turbo-surf-view/src/aria.rs
//! ARIA role + accessible-name helpers (port of `src/aria.mjs`) — pragmatic,
//! no-layout heuristics for getByRole/getByText resolution and the agent view.

use turbo_dom_parser::rtdom::Tree;
use turbo_dom_parser::rtdom::tree::Handle;

/// Collapsed, trimmed text content (the `dom-ops` `textOf` helper).
pub fn text_of(tree: &Tree, h: Handle) -> String {
    tree.text_content(h)
        .split_whitespace()
        .collect::<Vec<_>>()
        .join(" ")
}

fn input_role(type_attr: Option<&str>) -> &'static str {
    match type_attr {
        Some("checkbox") => "checkbox",
        Some("radio") => "radio",
        Some("button") | Some("submit") | Some("reset") => "button",
        _ => "textbox",
    }
}

/// Implicit ARIA role for a (lowercase) tag + optional input `type`.
pub fn implicit_role(tag: &str, type_attr: Option<&str>) -> &'static str {
    match tag {
        "input" => input_role(type_attr),
        "a" => "link",
        "button" => "button",
        "select" => "combobox",
        "textarea" => "textbox",
        _ => "generic",
    }
}

/// Resolved role: explicit `role` attribute, else the implicit tag/type role.
pub fn role_of(tree: &Tree, h: Handle) -> String {
    if let Some(explicit) = tree.get_attribute(h, "role") {
        return explicit.to_string();
    }
    let tag = tree.tag_name(h).unwrap_or_default().to_ascii_lowercase();
    let ty = tree
        .get_attribute(h, "type")
        .map(|s| s.to_ascii_lowercase());
    implicit_role(&tag, ty.as_deref()).to_string()
}

fn trimmed(s: Option<&str>) -> String {
    s.map(|v| v.trim().to_string()).unwrap_or_default()
}

/// Accessible name: aria-label > text > placeholder > value > title.
pub fn accessible_name(tree: &Tree, h: Handle) -> String {
    let candidates = [
        trimmed(tree.get_attribute(h, "aria-label")),
        tree.text_content(h).trim().to_string(),
        trimmed(tree.get_attribute(h, "placeholder")),
        trimmed(tree.get_attribute(h, "value")),
        trimmed(tree.get_attribute(h, "title")),
    ];
    candidates
        .into_iter()
        .find(|s| !s.is_empty())
        .unwrap_or_default()
}

fn id_list(tree: &Tree, h: Handle, attr: &str) -> Vec<String> {
    tree.get_attribute(h, attr)
        .unwrap_or("")
        .split_whitespace()
        .map(str::to_string)
        .collect()
}

fn resolve_ids(tree: &Tree, ids: &[String]) -> String {
    ids.iter()
        .filter_map(|id| tree.get_element_by_id(id))
        .map(|e| text_of(tree, e))
        .filter(|t| !t.is_empty())
        .collect::<Vec<_>>()
        .join(" ")
}

/// Accessible description: aria-describedby targets, else the `title` attribute.
pub fn accessible_description(tree: &Tree, h: Handle) -> String {
    let ids = id_list(tree, h, "aria-describedby");
    if !ids.is_empty() {
        return resolve_ids(tree, &ids);
    }
    trimmed(tree.get_attribute(h, "title"))
}

/// Accessible error message: aria-errormessage targets, only when aria-invalid.
pub fn accessible_error_message(tree: &Tree, h: Handle) -> String {
    if tree.get_attribute(h, "aria-invalid") != Some("true") {
        return String::new();
    }
    resolve_ids(tree, &id_list(tree, h, "aria-errormessage"))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn first(tree: &Tree, sel: &str) -> Handle {
        tree.query_selector(sel).unwrap()
    }

    #[test]
    fn implicit_roles() {
        assert_eq!(implicit_role("a", None), "link");
        assert_eq!(implicit_role("button", None), "button");
        assert_eq!(implicit_role("select", None), "combobox");
        assert_eq!(implicit_role("textarea", None), "textbox");
        assert_eq!(implicit_role("input", Some("checkbox")), "checkbox");
        assert_eq!(implicit_role("input", Some("submit")), "button");
        assert_eq!(implicit_role("input", None), "textbox");
        assert_eq!(implicit_role("div", None), "generic");
    }

    #[test]
    fn role_explicit_beats_implicit() {
        let tree = Tree::parse("<button role='tab'>x</button>");
        assert_eq!(role_of(&tree, first(&tree, "button")), "tab");
    }

    #[test]
    fn role_implicit_for_input_type() {
        let tree = Tree::parse("<input type='RADIO'>");
        assert_eq!(role_of(&tree, first(&tree, "input")), "radio");
    }

    #[test]
    fn accessible_name_priority() {
        let t1 = Tree::parse("<button aria-label='Lbl'>Text</button>");
        assert_eq!(accessible_name(&t1, first(&t1, "button")), "Lbl");
        let t2 = Tree::parse("<button>  Text  </button>");
        assert_eq!(accessible_name(&t2, first(&t2, "button")), "Text");
        let t3 = Tree::parse("<input placeholder='Ph'>");
        assert_eq!(accessible_name(&t3, first(&t3, "input")), "Ph");
        let t4 = Tree::parse("<input value='V'>");
        assert_eq!(accessible_name(&t4, first(&t4, "input")), "V");
        let t5 = Tree::parse("<a title='T'></a>");
        assert_eq!(accessible_name(&t5, first(&t5, "a")), "T");
        let t6 = Tree::parse("<span></span>");
        assert_eq!(accessible_name(&t6, first(&t6, "span")), "");
    }

    #[test]
    fn description_from_ids_then_title() {
        let t1 =
            Tree::parse("<input aria-describedby='d1 d2'><p id='d1'>one</p><p id='d2'>two</p>");
        assert_eq!(accessible_description(&t1, first(&t1, "input")), "one two");
        let t2 = Tree::parse("<input title='Tip'>");
        assert_eq!(accessible_description(&t2, first(&t2, "input")), "Tip");
    }

    #[test]
    fn error_message_only_when_invalid() {
        let ok = Tree::parse("<input aria-errormessage='e'><p id='e'>bad</p>");
        assert_eq!(accessible_error_message(&ok, first(&ok, "input")), "");
        let bad = Tree::parse("<input aria-invalid='true' aria-errormessage='e'><p id='e'>bad</p>");
        assert_eq!(accessible_error_message(&bad, first(&bad, "input")), "bad");
    }
}
